README.md
[![CI](https://github.com/GiantDarth/rbc_validator/actions/workflows/ci.yml/badge.svg)](https://github.com/GiantDarth/rbc_validator/actions/workflows/ci.yml)
[![License](https://img.shields.io/github/license/GiantDarth/rbc_validator)](https://opensource.org/licenses/Apache-2.0)

# rbc_validator
Primary and Auxiliary Authors:
* Christopher R. Philabaum  
* Christopher B. Coffey  
* Duane Booher

`rbc_validator` is a tool suite written as real-world implementations demonstrating
***RBC (Response-based Cryptography)***, both as a drop-in tool and as a benchmark set.
This solution does _not_ provide a real-world means of generating the disparate server-client
key pairs that need to be corrected (e.g. noisy sources, PUFs, etc.). Instead, the benchmark
functionality uses either pseudo-random number generation rather than direct true-random number
generation, or a targeted and known constant work by selectively choosing the "halfway point"
of a core or rank's workload.

## Requirements

### Hardware

The codebase relies on **SSE**, **AES-NI**, and support for _x86 Intrinsics_ `<x86intrin.h>`
(i.e., GCC 4.5+).

### Supported OSes

* Linux (64-bit)
* MacOS X Mojave or Later
* Windows 10 x64 (MSYS2 MinGW-w64)

### Dependencies

#### Linux

If on Debian-based distributions, the following command can be used to install all the requirements:
```bash
sudo apt-get install libomp-dev libopenmpi-dev openmpi-bin libgmp-dev libssl-dev
```

For other distributions, equivalents for _OpenMP (dev)_, _OpenMPI (bin, dev)_, _GMP (dev)_, and
_OpenSSL (dev)_ should be installed.

#### macOS

macOS implementations have been compiled and tested using Github Actions. This was done using Homebrew,
alongside Xcode. By default, MPI support is disabled on macoS, but can be manually enabled through
`-D MPI_ENABLED=On` in CMake and installing `open-mpi` in Homebrew.

This can be replicated by using the following command:
```zsh
brew install libomp gmp openssl@1.1
```

#### Windows 10 x64 (MSYS2 MinGW-w64)

Windows 10 x64 (MSYS MinGW-w64) implementations have been compiled and tested using Github Actions. This
was done using MSYS2. By default, MPI support is disabled on Windows and there's no clear path to get it
working.

This can be replicated by using the following commands:

First install [MSYS2](https://www.msys2.org/#installation) following the guide _verbatim_.
Then, open in MSYS MinGW 64-bit:
```bash
pacman -S --needed mingw-w64-x86_64-cmake gmp-devel openssl-devel
```

## Install

_rbc_validator_ uses CMake to best support cross-platform toolchains as much as possible. Windows is only
officially supported through MSYS2.

1. `mkdir build`
2. `cd build`
3. Either `ccmake ..` or `cmake ..` without curses (or `cmake -G "MSYS Makefiles" ..` for MSYS2).
4. `make`
5.
    1. Linux, macOS: `make install` (if you want to install to `/usr/local`)
    2. Windows: `./scripts/deploy_msys.sh` copies `rbc_validator` and the relevant DLL's to `dist/win_x64`

## Commands

All cryptographic algorithms are selected at runtime using the `--mode=` option, with the program being
split between an OpenMP implementation and an OpenMPI implementations. The _OpenMP_ implementations are
designed for **single machine** use, whereas the _OpenMPI_ ones target HPC platforms.

* `rbc_validator ...` (OpenMP)
* `mpirun rbc_validator_mpi ...` (MPI)

There are currently 4 algorithms supported:

* `rbc_validator --mode=none` (None, Key iteration only)
* `rbc_validator --mode=aes` (AES256)
* `rbc_validator --mode=chacha20` (ChaCha20)
* `rbc_validator --mode=ecc` (ECC Secp256r1)
* `rbc_validator --mode=md5`: (MD5)
* `rbc_validator --mode=sha1`: (SHA1)
* `rbc_validator --mode=sha224`: (SHA224)
* `rbc_validator --mode=sha256`: (SHA256)
* `rbc_validator --mode=sha384`: (SHA384)
* `rbc_validator --mode=sha512`: (SHA512)
* `rbc_validator --mode=sha3-224`: (SHA3-224)
* `rbc_validator --mode=sha3-256`: (SHA3-256)
* `rbc_validator --mode=sha3-384`: (SHA3-384)
* `rbc_validator --mode=sha3-512`: (SHA3-512)
* `rbc_validator --mode=kang12`: (KangrooTwelve, 32 bytes)


Some auxiliary commands also exist for testing the AES-256, ChaCha20, ECC-Secp256r1, and various
hash implementations against target keys and their associated ciphers:

* `aes256_test`
* `cipher_test`
* `ecc_test`
* `hash_test`

Finally, there exists a few Python scripts to generate some test data, as well as utility
functions.

## Usage

### Modes

#### Direct / Default

This is the default behavior of the main commands. Rather than as a tool for benchmarking,
"direct" mode requires that:

**None:**
1. The original server seed

**AES:**
1. The original server seed
2. The (potentially) corrupted client cipher
3. A shared message, in this case a 16 byte UUID

**ChaCha20:**
1. The original server seed
2. The (potentially) corrupted client cipher
3. A shared message, in this case a 16 byte UUID
4. A 16 byte IV (a 12 byte nonce prepended with 4 bytes of 0's)

**ECC:**
1. The original server seed
2. The (potentially) corrupted client public key (in compressed or uncompressed form)

**Hash:**
1. The original server private key
2. The (potentially) corrupted client digest

#### Random

Instead of giving the above parameters, the program will simulate the corruption by first 
generating a random server key, corrupting it uniformly randomly by the given _mismatches_,
and either generating the disparate ciphers (AES), or disparate public-private key pairs (ECC).

#### Benchmark

Since the range of possible steps a core / rank might need to take ranges extremely wildly
between _0_ and _256 choose_ m, getting useful benchmark is limited by the amount of runs
it would take to get a strong statistical average. This is compounded by the fact that as
_mismatches_ goes up, the time taken is exponential no matter the amount of compute power
thrown at it.

For this reason, _Benchmark mode_ was made. This made behaves similarly to _Random mode_,
but instead of uniformly randomly choosing the location of the mismatches, it uniformly
randomly chooses the core / rank. Then, whichever is chosen, the location of the mismatches
is strategically chosen to be exactly half of the core's / rank's sequentially load.

### Arguments

The main commands have required parameters when not in `--random` nor `--benchmark` mode.

* `HOST_SEED`: The 32 byte host seed to start from, in hexadecimal.
* `CLIENT_CIPHER`: An _n_ byte cipher produced by the client, in hexadecimal. _n_ is 32 bytes for
  `--mode=aes` (specifically AES256) and 32 bytes for `--mode=chacha20`.
* `UUID`: A UUID in canonical form (hhhhhhhh-hhhh-hhhh-hhhh-hhhhhhhhhhhh) that's shared between
  client and host when using `--mode=[aes,chacha20]`.
* `UUID`: An _n_ byte UUID IV that's shared between client and host when using
  `--mode=[chacha20]`. _n_ is 16 bytes for `--mode=chacha20`
* `CLIENT_PUB_KEY`: An _n_ bytes public key produced by the client, in hexadecimal. For
  `--mode=ecc` (specifically Secp256r1), `CLIENT_PUB_KEY` can either be in compressed form
  (_n_ = 33 bytes) or uncompressed form (_n_ = 65 bytes).
* `CLIENT_DIGEST`: An _n_-bit digest produced by the client, in hexadecimal. The size of _n_ is
  required the following hash functions:
  * `--mode=md5`: _n_ = 16 bytes
  * `--mode=sha1`: _n_ = 20 bytes
  * `--mode=sha224`: _n_ = 24 bytes
  * `--mode=sha256`: _n_ = 32 bytes
  * `--mode=sha384`: _n_ = 48 bytes
  * `--mode=sha512`: _n_ = 64 bytes
  * `--mode=sha3-224`: _n_ = 24 bytes
  * `--mode=sha3-256`: _n_ = 32 bytes
  * `--mode=sha3-384`: _n_ = 48 bytes
  * `--mode=sha3-512`: _n_ = 64 bytes
  * `--mode=kang12`: _n_ = 32 bytes

All the main commands have (mostly) the same options:

* `--usage`: An auto-generated simple usage message made by Argp, which is the same if an
  invalid argument is used.
* `-?, --help`: The main source of information on how to use each command, the arguments list,
  their use, default values, and their ranges.
* `--mode=[none,aes,chacha20,ecc,md5,sha1,sha224,sha256,sha384,sha512,sha3-224,sha3-256,sha3-384,sha3-512,kang12]`:
  The only required option; necessary to decide which cryptographic function to use.
* `-m, --mismatches=value`: Give the maximum range of hamming distance / errors to test up to
  and including. If not given, then the maximum range is the size of the key in bits.
* `-a, --all`: Ignore any early exit method and have each core / rank carry on through its
  workload exhaustively.
* `-f, --fixed`: Only test the given _m_, rather than doing the more realistic checks of
  _0, 1, 2, ..., m_.
* `-s, --subkey=value`: The size (in bits) of the key fragment / subkey if using key
  fragmentation. Only up to the first _s_ bits will be manipulated.
* `-b, --benchmark`: Forcefully pick the corrupted key to be exactly halfway through a
  randomly chosen core's / rank's workload. Mutually exclusive with default and `-r`/`--random`
  mode.
* `-r, --random`: Randomly pick the corrupted key along a "number line" from _0_ to
_ 256 choose m_.
* `-c, --count`: Count each corrupted key generated and tested and display at the end.
* `-v, --verbose`: Produce verbose and benchmarking output to _stderr_. Otherwise, only the
  found key is printed to _stdout_.
* `-V, --version`: Print the program version.

The OpenMP implementations also allow direct control of the thread count used:

* `-t, --threads=count`: The number of the threads to use, which defaults to the number of
  threads reported on the machine.


CHANGELOG.md
# rbc_validator Changelog

## 1.0.0 (May 21, 2021)

### Features

* Added support for the following symmetric keys (ciphers):
  * AES-256-ECB
  * ChaCha20
* Added support for the following public keys:
  * ECC Secp256r1
* Added support for the following hash functions:
  * MD5
  * SHA1
  * SHA224
  * SHA256
  * SHA384
  * SHA512
  * SHA3-224
  * SHA3-256
  * SHA3-384
  * SHA3-512
  * SHAKE128
  * SHAKE256
  * KangarooTwelve
* Added support for the following modes:
  * **Direct**: Pass in a known host seed and known client output to derive the client's seed
  * `--random`: Generate the seed and errors randomly. Useful for benchmarking that models
    uniformly random distributions.
  * `--benchmark`: Generate the seed and errors such that the errors are exactly halfway through a
    core's/rank's workload. Useful for benchmarking but you're short on time or want
    reproducibility.
* Added support for both OpenMP and OpenMPI implementations
* Added the following options:
  * `--all`: Enforces search to continue through an entire Hamming distance's search space even if a
    matching seed is found
  * `--fixed`: _Only_ searches a given Hamming distance's search space (no more nor less)
* Added the ability to choose between OpenSSL vs. a faster custom AES256 implementation using
  `ALWAYS_EVP_AES`. Defaults to the custom implementation
* Added the ability to choose between EVP vs. low level OpenSSL implementation of MD5, SHA1, and
  SHA2 using `ALWAYS_EVP_HASH`. Defaults to the low level implementations.
* Added the ability to choose between OpenSSL vs. XKCP implementations for SHA3 and SHAKE
  using `ALWAYS_EVP_SHA3`. Defaults to OpenSSL implementations. _(Note: Testing has found that the
  OpenSSL tends to be faster than the `generic64`, `SSSE3`, and `AVX` XKCP compilations on Zen 2.
  `AVX2` is about on par. `AVX512` XKCP is significantly faster than OpenSSL on Xeon Skylake
  CPUs.)_
* Added the ability to manually select the architecture to build XKCP against between:
  * `reference`
  * `reference32bits`
  * `compact`
  * `generic32`
  * `generc32lc`
  * `generic64`
  * `generic64lc`
  * `SSSE3`
  * `AVX`
  * `AVX2`
  * `AVX512`
  
  Defaults to `generic64` to maximize compatibility while balancing for basic optimizations.
  

src/uECC.h
/* Copyright 2014, Kenneth MacKay. Licensed under the BSD 2-clause license. */

#ifndef _UECC_H_
#define _UECC_H_

#include <stdint.h>

/* Platform selection options.
If uECC_PLATFORM is not defined, the code will try to guess it based on compiler macros.
Possible values for uECC_PLATFORM are defined below: */
#define uECC_arch_other 0
#define uECC_x86        1
#define uECC_x86_64     2
#define uECC_arm        3
#define uECC_arm_thumb  4
#define uECC_arm_thumb2 5
#define uECC_arm64      6
#define uECC_avr        7

/* If desired, you can define uECC_WORD_SIZE as appropriate for your platform (1, 4, or 8 bytes).
If uECC_WORD_SIZE is not explicitly defined then it will be automatically set based on your
platform. */

/* Optimization level; trade speed for code size.
   Larger values produce code that is faster but larger.
   Currently supported values are 0 - 4; 0 is unusably slow for most applications.
   Optimization level 4 currently only has an effect ARM platforms where more than one
   curve is enabled. */
#ifndef uECC_OPTIMIZATION_LEVEL
#define uECC_OPTIMIZATION_LEVEL 2
#endif

/* uECC_SQUARE_FUNC - If enabled (defined as nonzero), this will cause a specific function to be
used for (scalar) squaring instead of the generic multiplication function. This can make things
faster somewhat faster, but increases the code size. */
#ifndef uECC_SQUARE_FUNC
#define uECC_SQUARE_FUNC 0
#endif

/* uECC_VLI_NATIVE_LITTLE_ENDIAN - If enabled (defined as nonzero), this will switch to native
little-endian format for *all* arrays passed in and out of the public API. This includes public
and private keys, shared secrets, signatures and message hashes.
Using this switch reduces the amount of call stack memory used by uECC, since less intermediate
translations are required.
Note that this will *only* work on native little-endian processors and it will treat the uint8_t
arrays passed into the public API as word arrays, therefore requiring the provided byte arrays
to be word aligned on architectures that do not support unaligned accesses.
IMPORTANT: Keys and signatures generated with uECC_VLI_NATIVE_LITTLE_ENDIAN=1 are incompatible
with keys and signatures generated with uECC_VLI_NATIVE_LITTLE_ENDIAN=0; all parties must use
the same endianness. */
#ifndef uECC_VLI_NATIVE_LITTLE_ENDIAN
#define uECC_VLI_NATIVE_LITTLE_ENDIAN 0
#endif

/* Curve support selection. Set to 0 to remove that curve. */
#ifndef uECC_SUPPORTS_secp160r1
#define uECC_SUPPORTS_secp160r1 1
#endif
#ifndef uECC_SUPPORTS_secp192r1
#define uECC_SUPPORTS_secp192r1 1
#endif
#ifndef uECC_SUPPORTS_secp224r1
#define uECC_SUPPORTS_secp224r1 1
#endif
#ifndef uECC_SUPPORTS_secp256r1
#define uECC_SUPPORTS_secp256r1 1
#endif
#ifndef uECC_SUPPORTS_secp256k1
#define uECC_SUPPORTS_secp256k1 1
#endif

/* Specifies whether compressed point format is supported.
   Set to 0 to disable point compression/decompression functions. */
#ifndef uECC_SUPPORT_COMPRESSED_POINT
#define uECC_SUPPORT_COMPRESSED_POINT 1
#endif

struct uECC_Curve_t;
typedef const struct uECC_Curve_t * uECC_Curve;

#ifdef __cplusplus
extern "C"
{
#endif

#if uECC_SUPPORTS_secp160r1
uECC_Curve uECC_secp160r1(void);
#endif
#if uECC_SUPPORTS_secp192r1
uECC_Curve uECC_secp192r1(void);
#endif
#if uECC_SUPPORTS_secp224r1
uECC_Curve uECC_secp224r1(void);
#endif
#if uECC_SUPPORTS_secp256r1
uECC_Curve uECC_secp256r1(void);
#endif
#if uECC_SUPPORTS_secp256k1
uECC_Curve uECC_secp256k1(void);
#endif

/* uECC_RNG_Function type
The RNG function should fill 'size' random bytes into 'dest'. It should return 1 if
'dest' was filled with random data, or 0 if the random data could not be generated.
The filled-in values should be either truly random, or from a cryptographically-secure PRNG.

A correctly functioning RNG function must be set (using uECC_set_rng()) before calling
uECC_make_key() or uECC_sign().

Setting a correctly functioning RNG function improves the resistance to side-channel attacks
for uECC_shared_secret() and uECC_sign_deterministic().

A correct RNG function is set by default when building for Windows, Linux, or OS X.
If you are building on another POSIX-compliant system that supports /dev/random or /dev/urandom,
you can define uECC_POSIX to use the predefined RNG. For embedded platforms there is no predefined
RNG function; you must provide your own.
*/
typedef int (*uECC_RNG_Function)(uint8_t *dest, unsigned size);

/* uECC_set_rng() function.
Set the function that will be used to generate random bytes. The RNG function should
return 1 if the random data was generated, or 0 if the random data could not be generated.

On platforms where there is no predefined RNG function (eg embedded platforms), this must
be called before uECC_make_key() or uECC_sign() are used.

Inputs:
    rng_function - The function that will be used to generate random bytes.
*/
void uECC_set_rng(uECC_RNG_Function rng_function);

/* uECC_get_rng() function.

Returns the function that will be used to generate random bytes.
*/
uECC_RNG_Function uECC_get_rng(void);

/* uECC_curve_private_key_size() function.

Returns the size of a private key for the curve in bytes.
*/
int uECC_curve_private_key_size(uECC_Curve curve);

/* uECC_curve_public_key_size() function.

Returns the size of a public key for the curve in bytes.
*/
int uECC_curve_public_key_size(uECC_Curve curve);

/* uECC_make_key() function.
Create a public/private key pair.

Outputs:
    public_key  - Will be filled in with the public key. Must be at least 2 * the curve size
                  (in bytes) long. For example, if the curve is secp256r1, public_key must be 64
                  bytes long.
    private_key - Will be filled in with the private key. Must be as long as the curve order; this
                  is typically the same as the curve size, except for secp160r1. For example, if the
                  curve is secp256r1, private_key must be 32 bytes long.

                  For secp160r1, private_key must be 21 bytes long! Note that the first byte will
                  almost always be 0 (there is about a 1 in 2^80 chance of it being non-zero).

Returns 1 if the key pair was generated successfully, 0 if an error occurred.
*/
int uECC_make_key(uint8_t *public_key, uint8_t *private_key, uECC_Curve curve);

/* uECC_shared_secret() function.
Compute a shared secret given your secret key and someone else's public key.
Note: It is recommended that you hash the result of uECC_shared_secret() before using it for
symmetric encryption or HMAC.

Inputs:
    public_key  - The public key of the remote party.
    private_key - Your private key.

Outputs:
    secret - Will be filled in with the shared secret value. Must be the same size as the
             curve size; for example, if the curve is secp256r1, secret must be 32 bytes long.

Returns 1 if the shared secret was generated successfully, 0 if an error occurred.
*/
int uECC_shared_secret(const uint8_t *public_key,
                       const uint8_t *private_key,
                       uint8_t *secret,
                       uECC_Curve curve);

#if uECC_SUPPORT_COMPRESSED_POINT
/* uECC_compress() function.
Compress a public key.

Inputs:
    public_key - The public key to compress.

Outputs:
    compressed - Will be filled in with the compressed public key. Must be at least
                 (curve size + 1) bytes long; for example, if the curve is secp256r1,
                 compressed must be 33 bytes long.
*/
void uECC_compress(const uint8_t *public_key, uint8_t *compressed, uECC_Curve curve);

/* uECC_decompress() function.
Decompress a compressed public key.

Inputs:
    compressed - The compressed public key.

Outputs:
    public_key - Will be filled in with the decompressed public key.
*/
void uECC_decompress(const uint8_t *compressed, uint8_t *public_key, uECC_Curve curve);
#endif /* uECC_SUPPORT_COMPRESSED_POINT */

/* uECC_valid_public_key() function.
Check to see if a public key is valid.

Note that you are not required to check for a valid public key before using any other uECC
functions. However, you may wish to avoid spending CPU time computing a shared secret or
verifying a signature using an invalid public key.

Inputs:
    public_key - The public key to check.

Returns 1 if the public key is valid, 0 if it is invalid.
*/
int uECC_valid_public_key(const uint8_t *public_key, uECC_Curve curve);

/* uECC_compute_public_key() function.
Compute the corresponding public key for a private key.

Inputs:
    private_key - The private key to compute the public key for

Outputs:
    public_key - Will be filled in with the corresponding public key

Returns 1 if the key was computed successfully, 0 if an error occurred.
*/
int uECC_compute_public_key(const uint8_t *private_key, uint8_t *public_key, uECC_Curve curve);

/* uECC_sign() function.
Generate an ECDSA signature for a given hash value.

Usage: Compute a hash of the data you wish to sign (SHA-2 is recommended) and pass it in to
this function along with your private key.

Inputs:
    private_key  - Your private key.
    message_hash - The hash of the message to sign.
    hash_size    - The size of message_hash in bytes.

Outputs:
    signature - Will be filled in with the signature value. Must be at least 2 * curve size long.
                For example, if the curve is secp256r1, signature must be 64 bytes long.

Returns 1 if the signature generated successfully, 0 if an error occurred.
*/
int uECC_sign(const uint8_t *private_key,
              const uint8_t *message_hash,
              unsigned hash_size,
              uint8_t *signature,
              uECC_Curve curve);

/* uECC_HashContext structure.
This is used to pass in an arbitrary hash function to uECC_sign_deterministic().
The structure will be used for multiple hash computations; each time a new hash
is computed, init_hash() will be called, followed by one or more calls to
update_hash(), and finally a call to finish_hash() to produce the resulting hash.

The intention is that you will create a structure that includes uECC_HashContext
followed by any hash-specific data. For example:

typedef struct SHA256_HashContext {
    uECC_HashContext uECC;
    SHA256_CTX ctx;
} SHA256_HashContext;

void init_SHA256(uECC_HashContext *base) {
    SHA256_HashContext *context = (SHA256_HashContext *)base;
    SHA256_Init(&context->ctx);
}

void update_SHA256(uECC_HashContext *base,
                   const uint8_t *message,
                   unsigned message_size) {
    SHA256_HashContext *context = (SHA256_HashContext *)base;
    SHA256_Update(&context->ctx, message, message_size);
}

void finish_SHA256(uECC_HashContext *base, uint8_t *hash_result) {
    SHA256_HashContext *context = (SHA256_HashContext *)base;
    SHA256_Final(hash_result, &context->ctx);
}

... when signing ...
{
    uint8_t tmp[32 + 32 + 64];
    SHA256_HashContext ctx = {{&init_SHA256, &update_SHA256, &finish_SHA256, 64, 32, tmp}};
    uECC_sign_deterministic(key, message_hash, &ctx.uECC, signature);
}
*/
typedef struct uECC_HashContext {
    void (*init_hash)(const struct uECC_HashContext *context);
    void (*update_hash)(const struct uECC_HashContext *context,
                        const uint8_t *message,
                        unsigned message_size);
    void (*finish_hash)(const struct uECC_HashContext *context, uint8_t *hash_result);
    unsigned block_size; /* Hash function block size in bytes, eg 64 for SHA-256. */
    unsigned result_size; /* Hash function result size in bytes, eg 32 for SHA-256. */
    uint8_t *tmp; /* Must point to a buffer of at least (2 * result_size + block_size) bytes. */
} uECC_HashContext;

/* uECC_sign_deterministic() function.
Generate an ECDSA signature for a given hash value, using a deterministic algorithm
(see RFC 6979). You do not need to set the RNG using uECC_set_rng() before calling
this function; however, if the RNG is defined it will improve resistance to side-channel
attacks.

Usage: Compute a hash of the data you wish to sign (SHA-2 is recommended) and pass it to
this function along with your private key and a hash context. Note that the message_hash
does not need to be computed with the same hash function used by hash_context.

Inputs:
    private_key  - Your private key.
    message_hash - The hash of the message to sign.
    hash_size    - The size of message_hash in bytes.
    hash_context - A hash context to use.

Outputs:
    signature - Will be filled in with the signature value.

Returns 1 if the signature generated successfully, 0 if an error occurred.
*/
int uECC_sign_deterministic(const uint8_t *private_key,
                            const uint8_t *message_hash,
                            unsigned hash_size,
                            const uECC_HashContext *hash_context,
                            uint8_t *signature,
                            uECC_Curve curve);

/* uECC_verify() function.
Verify an ECDSA signature.

Usage: Compute the hash of the signed data using the same hash as the signer and
pass it to this function along with the signer's public key and the signature values (r and s).

Inputs:
    public_key   - The signer's public key.
    message_hash - The hash of the signed data.
    hash_size    - The size of message_hash in bytes.
    signature    - The signature value.

Returns 1 if the signature is valid, 0 if it is invalid.
*/
int uECC_verify(const uint8_t *public_key,
                const uint8_t *message_hash,
                unsigned hash_size,
                const uint8_t *signature,
                uECC_Curve curve);

#ifdef __cplusplus
} /* end of extern "C" */
#endif

#endif /* _UECC_H_ */
